pkg/controller/pkg/revision/establisher.go

```diff
@@ -131,30 +133,42 @@ func (e *APIEstablisher) Establish(ctx context.Context, objs []runtime.Object, p
 	return resourceRefs, nil
 }
 
-func (e *APIEstablisher) create(ctx context.Context, obj resource.Object, parent resource.Object, control bool, opts ...client.CreateOption) error {
+func (e *APIEstablisher) create(ctx context.Context, obj runtime.Object, parent resource.Object, control bool, opts ...client.CreateOption) error {
+	o := obj.DeepCopyObject().(resource.Object)
 	ref := meta.AsController(meta.TypedReferenceTo(parent, parent.GetObjectKind().GroupVersionKind()))
 	if !control {
 		ref = meta.AsOwner(meta.TypedReferenceTo(parent, parent.GetObjectKind().GroupVersionKind()))
 	}
 	// Overwrite any owner references on the desired object.
-	obj.SetOwnerReferences([]metav1.OwnerReference{ref})
+	o.SetOwnerReferences([]metav1.OwnerReference{ref})
 	return e.client.Create(ctx, obj, opts...)
 }
 
-func (e *APIEstablisher) update(ctx context.Context, current, desired resource.Object, parent resource.Object, control bool, opts ...client.UpdateOption) error {
+func (e *APIEstablisher) update(ctx context.Context, current, desired runtime.Object, parent resource.Object, control bool, opts ...client.UpdateOption) error {
+	c := current.DeepCopyObject().(resource.Object)
+	d := desired.DeepCopyObject().(resource.Object)
 	if !control {
-		meta.AddOwnerReference(current, meta.AsOwner(meta.TypedReferenceTo(parent, parent.GetObjectKind().GroupVersionKind())))
-		return e.client.Update(ctx, current, opts...)
+		ownerRef := meta.AsOwner(meta.TypedReferenceTo(parent, parent.GetObjectKind().GroupVersionKind()))
+		for _, ref := range c.GetOwnerReferences() {
+			if ref == ownerRef {
```

Comment:
Will struct equality work as expected when the struct has pointer fields? (e.g. `ref.Controller`)?

Reply:
Also, this logic looks pretty similar to what `meta.AddOwnerReference` does. Presumably we could also copy the original, then call `meta.AddOwnerReference`, then use `cmp` to determine whether they changed, but I suspect your current implementation would be more efficient.

Reply:
> Will struct equality work as expected when the struct has pointer fields?

This comparison is only used on owners, never controllers, so we don't actually end up comparing any pointers that are not `nil`.